add parallel blog - epsode 1 (#220)
ARCLeeds/arcleeds.github.io: blog/_posts/2023-10-19-parallelSeriesEpisode1.md

---
layout: page

> title:  Exploring Parallel Python Series - Episode 1

Author: ```suggestion
title:  A Journey into Parallel Python - Episode 1
```

- With 10^7 numbers, the execution time reduced to approximately 18.56 seconds, achieving a speedup of about 3.82 times.
- For 10^8 numbers, we observed a substantial reduction in execution time, with approximately 4 minutes (238.96 seconds), making it approximately 8.14 times faster than the serial approach.

Although we have attempted to divide the workload into 40 cores theoretically, the improvement in simulation time is not significant. This may be due to the nature of the problem as checking whether a number is prime or not is not a time-consuming task. The lower performance improvement can be attributed to the substantial overhead introduced by parallelism, particularly the creation and management of multiple processes and communication between them. Often, the speed at which processes complete their tasks can surpass the efficiency of communication between cores. This leads to underutilization of available CPU cores. The image below depicts this finding: we never managed to have 40 cores using 100% of CPU for Python tasks.

![Plain Parallel CPU Usage](/images/blog/parallelSeries/episode1/cpu_plain.png)

#### Parallel Chunk Approach

> Next, we implemented a parallel chunk approach, where we divided the range of numbers into chunks and assigned each chunk to a separate process. This is an attempt to create larger process to avoid the communication overhead. The task here is find the right ballance between the time we loose managing all process and keeping all cores using 100% of CPU as much as possible. This approach allowed us to fine-tune the level of parallelism. The results for 40 chunks were particularly impressive:

Reviewer: Time we *lose managing all processes*

Reviewer: ```suggestion
Next, we implemented a parallel chunk approach, where we divided the range of numbers into chunks and assigned each chunk to a separate process. This is an attempt to create larger process to avoid the communication overhead. The task here is find the right ballance between the time we lose managing all processes and keeping all cores using 100% of CPU as much as possible. This approach allowed us to fine-tune the level of parallelism. The results for 40 chunks were particularly impressive:
```

- For 10^6 numbers, the parallel chunk approach took approximately 0.27 seconds, showcasing a substantial improvement, approximately 10.67 times faster than the serial version.
- With 10^7 numbers, execution time reduced to approximately 4.09 seconds, achieving a remarkable speedup of about 17.38 times.
- Even with the most extensive range of 10^8 numbers, the execution time was significantly reduced to approximately 1.62 minutes (97.32 seconds), making it approximately 19.98 times faster than the serial approach.

By using chunks, every process spend more time running and less time communicating. This improves the overall efficiency of the experiment. As you can see bellow, with this approach we manage to hit all 40 cores:

![Chunk Parallel CPU Usage](/images/blog/parallelSeries/episode1/cpu_chunk.png)

> You can notice that 40 chunks is not the best option. This happens because check the interval 1-10 takes much less time than check the range 100-110, for example.
> So at some point some cores will finish where others are going to be working at maximum capacity. The idea is to find a good balance where you have enough numbers of chunks to when the quick ranges finish, a new process can start in that core, but not so many chunks that your overhead comunication is going to use all the time you won with the cores otimization

Author: ```suggestion
```
